File: lesson11/TagLayout/app/src/main/java/com/chinalwb/taglayout/SquareImageView.java
package com.chinalwb.taglayout;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatImageView;
import android.util.AttributeSet;
import android.widget.ImageView;

public class SquareImageView extends AppCompatImageView {
    public SquareImageView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }


    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
         super.onMeasure(widthMeasureSpec, heightMeasureSpec);

//        int width = MeasureSpec.getSize(widthMeasureSpec);
//        int height = MeasureSpec.getSize(heightMeasureSpec);
//        int size = Math.max(width, height);
//        setMeasuredDimension(size, size);
    }

    @Override
    public void layout(int l, int t, int r, int b) {
        int width = r - l;
        int height = b - t;
        int size = Math.max(width, height);
//        super.layout(l,t,r,b);
         super.layout(l, t, l + size, t + size);
    }
}

File: lesson11/TagLayout/app/src/main/java/com/chinalwb/taglayout/v2/SquareImageView.kt
package com.chinalwb.taglayout.v2

import android.content.Context
import android.graphics.Canvas
import android.util.AttributeSet
import android.util.Log
import android.widget.ImageView

class SquareImageView (context: Context, attributeSet: AttributeSet): ImageView(context,
        attributeSet) {

    private var ow = 0
    private var oh = 0

    var changed = false
    public fun changeSize() {
        if (ow > oh) {
            oh = ow
        } else {
            ow = oh
        }

        changed = true
        Log.w("XX", "Changed size!! $ow , $oh")
        requestLayout()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        Log.w("XX", "on size changed !! new width: $w, $h -- old width $oldw, old height $oldh")
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        var mw = MeasureSpec.getSize(widthMeasureSpec)
        var mh = MeasureSpec.getSize(heightMeasureSpec)
        if (ow == 0) {
            ow = mw
        }
        if (oh == 0) {
            oh = mh
        }

        if (mw > mh) {
            mw = mh
        } else {
            mh = mw
        }

        if (changed) {
            setMeasuredDimension(ow, oh)
            return
        }
        setMeasuredDimension(mw, mh)
        Log.w("XX", "first time: $mw $mh")
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        Log.w("XX", "ondraw !! $changed")
    }
}